lib/IDE/Refactoring.cpp

```diff
@@ -5373,28 +5583,54 @@ class AsyncConverter : private SourceEntityWalker {
     }
   }
 
-  // TODO: Check for clashes with existing names and add all decls, not just
-  //       params
+  /// Add a mapping from each passed parameter to a new name, possibly
+  /// synthesizing a new one if hoisting it would cause a redeclaration or
+  /// shadowing. If there's no bound name and \c AddIfMissing is false, no
+  /// name will be added.
   void prepareNames(const ClassifiedBlock &Block,
                     ArrayRef<const ParamDecl *> Params,
                     bool AddIfMissing = true) {
-    for (auto *Param : Params) {
-      StringRef Name = Block.boundName(Param);
-      if (!Name.empty()) {
-        Names[Param] = Name.str();
-        continue;
-      }
+    for (auto *PD : Params) {
+      StringRef Name = Block.boundName(PD);
+      if (Name.empty() && AddIfMissing)
+        Name = PD->getNameStr();
+      assignName(PD, Name);
+    }
 
-      if (!AddIfMissing)
-        continue;
+    for (auto &Entry : Block.aliases()) {
+      auto It = Names.find(Entry.second);
+      assert(It != Names.end() && "Param should already have an entry");
+      Names[Entry.first] = It->second;
+    }
+  }
 
-      auto ParamName = Param->getNameStr();
-      if (ParamName.startswith("$")) {
-        Names[Param] = "val" + ParamName.drop_front().str();
-      } else {
-        Names[Param] = ParamName.str();
-      }
+  StringRef assignName(const Decl *D, StringRef Name) {
```

Comment:
Couldn’t the `Decl *` parameter be a `VarDecl *`? You are always only passing in `VarDecl` or `ParamDecl` (which is also a `VarDecl`).  Then you also wouldn’t need the `Name` parameter, but could use `getNameStr()` on the `VarDecl`. And since you now have access to the `Identifier` of the decl, I *believe* you could also replace the check for `""` and `"_"` below by `VD->getName().empty()`. And you could get rid if the `dyn_cast<VarDecl>`

Once you changed that, I think the more accurate name of this function would be `getUniqueName`. 

And: Could you add a doc-comment? After my proposed changes, I’d suggest
```
Returns a unique name for the variable declared by \p VD that doesn't clash with any hoisted variables.
```

Reply:
The `Name` parameter is primarily for the already-bound-name case, ie. a name coming from a let binding in a condition. The VarDecl* change, rename, and doc comment still make sense though - thanks.

Reply:
Ah, I see. Missed the case where `Name` isn’t `VD->getNameStr()`